### ISSUE TITLE
[question] couldn't we let nestjs controller methods decide the return types?

### ISSUE THREAD

**Issue by renatoaraujoc** [ts-rest/ts-rest#197]
Hello team,

Would it be possible for the NestJS integration to let the endpoints themselves define their return types?

We'd make it force a base return type shape containing the http status code like:

```ts
type TsRestResponseBase = { [Status in 200 | 201]: unknown } // this is just a silly example

@Controller()
export class PostController implements NestControllerInterface<typeof c> {
  constructor(private readonly postService: PostService) {}

  @TsRest(c.getPost)
  async getPost(@TsRestRequest() { params: { id } }: TsRestResponseBase {
    const post = await this.postService.getPost(id);

    if (!post) {
      return { status: 404 as const, body: null };
    }

    return { status: 200 as const, body: post };
  }
}

export const finalPostControllerRoutes = resolveNestJsController<PostController>(); // another silly example
```

And somehow the router/contract infer the final return type from that end point **dinamically**?

Or would this idea be worse than the actual behavior of statically forcing all the return types of controller methods thru the contract?

Thank you!

**Comment by Gabrola** [comment]
Hey @renatoaraujoc! While I think it's technically possible, I think that would be a huge departure from ts-rest's concept as we'd no longer have the concept of a _contract_, and we'd also lose the ability to use validateable Zod schemas for response types. I personally believe it's safer to explicitly type response types in a contract and then make sure the API's conforms to that, as that way you can ensure that you never expose more information accidentally, especially with response validation enabled which would not be possible without Zod on the responses.

This is also enables teams that have separate backend and frontend engineers to work faster, since you can define the contract and then start working immediately after that. It would also act as an always up-to-date source of truth and documentation.

**Comment by renatoaraujoc** [comment]
Hey @Gabrola, thanks for your input.

I understand you point, my question arises because this is technically how tRPC works nowadays, afaik you write queries and mutations and the return type is automatically inferred and reflected on the frontend.

Do we really need to validate responses when they they can be inferred? Maybe that could be a choice if one chooses the stronger contract way, or ts-rest api could provide a second way which follows tRPC style and let the frontend handle the queries and mutations data response changes.

I work with nestjs and prisma, so imagine I have to generate zod types for each prisma domain (okay I can use a "toZod" sort of thing), slice then apart and construct a response combined of many sliced prisma domain objects (incl. relations), seems a bit overwhelming... this would be automatically be solved with inferred return types and leave the frontend handle the data (and breaking changes) the way they have to.

What do you think? 

Renato

**Comment by Gabrola** [comment]
@renatoaraujoc Yes, that would be exactly what tRPC would be doing. Other than the protocol, I think there's too much of an overlap between a tRPC-style ts-rest and tRPC itself to justify creating this solution, but I might be wrong though.

I'm genuinely keen to hear your thoughts on why you'd like to use ts-rest in this case then rather than tRPC outright? Is it just because tRPC doesn't have a NestJS adapter?

**Comment by renatoaraujoc** [comment]
@Gabrola ts-rest showed up as an alternative exactly when trying to find an integrated solution from tRPC and NestJS. Found about this package in an issue in their own repo.

The problem nowadays is that you need absurd amount of packages and configuration to achieve the type sharing between a frontend and backend (graphql + generators, hacks to infer return types or construct shared return types which are very bad to maintain). Okay, tRPC solves all of this already but they will not provide a self maintained solution to integrate with NestJS.

I saw the adapter you're talking about but digging thru the source code it has shown to basically be a "hack" to inject a middleware and route the responses to a tRPC "module" while being able to inject services from NestJS.

Not only this solution diverges from how NestJS is supposed to work but also it's not an official integration, this shouldn't be maintained in the long run.

To summarize, your solution with ts-rest integrated very nicely with NestJS, providing tRPC-like experience while also leaving the framework functioning like it's supposed to be. 

This is my feedback and if you think it's viable to provide a "let ts-rest infer types from the backend and send them to frontend", it'd be amazing 🙂 

Renato

**Comment by renatoaraujoc** [comment]
Hey @Gabrola,

> For example, if you want add a new field to a database entity and change all endpoints which return this entity in their responses, your first step would be to determine which endpoints return this entity. Having this defined in the contracts themselves makes this a very trivial task. The next step would be to add this field in the contracts; a change that is very quick and only localized to one, or a small number of tiny files. After that a simple TS type-check command would spit out errors clearly pointing towards the backend code that you need to change, making that also probably a quick change (say just add a new select field in a prisma query).
> 
> In the case of inferring response types, you'd have to dig through controllers and services to even determine which endpoints are returning this database entity, which is a much more tedious process if you have a large backend. After this, you run the risk of changing a service that is used by other controllers/endpoints and unintendedly exposing extra and potentially sensitive data in another endpoint. With explicit return types, and response validation, you wouldn't have this risk.

You've beaten my arguments here. Needless to say anything else, 100% agree. Building contracts can be tedius at first but provide much smoother way for maintenance and build of new features.

Thanks for your time for explaining all of this! I'll close this.

Renato
